### PR TITLE
Spec sync

### PR DESCRIPTION
**Pull request by ralexstokes**
### What was wrong?

I was looking at exactly what functionality remained to reach our first pass at a full state transition function for the beacon chain.

### How was it fixed?

The existing code was compared to the current spec. There was some refactoring done to make parts of the state transition more manageable.

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

**Comment by ralexstokes** [comment]
need to fix linter

**Comment by ralexstokes** [comment]
@hwwhww heads up if this is merged, there *may* be a merge conflict w/ #314.

File: eth2/beacon/state_machines/forks/serenity/state_transitions.py [hunk 1]
```diff
@@ -98,12 +98,12 @@ def per_block_transition(self,
         state = process_attestations(state, block, self.config)
         # TODO: state = process_deposits(state, block, self.config)
         # TODO: state = process_voluntary_exits(state, block, self.config)
-        # TODO: validate_custody(state, block, self.config)
```
Comment (ralexstokes) [inline eth2/beacon/state_machines/forks/serenity/state_transitions.py]: @hwwhww i don't see any explicit custody logic in the current spec for the top-level state transition here. lmk if we should leave this `TODO` or it is safe to remove.

**Comment by ralexstokes** [comment]
pending spec changes, here is the missing functionality required to complete the beacon chain state transition function:

process_randao -- ralexstokes will enable, addressing issue #284
process_attester_slashings -- PR #299
process_deposits -- issue #238 (open, check w/ hwwhww)
process_voluntary_exits -- issue #238 (open, check w/ hwwhww)
process_transfers -- issue #325 (open)
process_eth1_data_votes -- issue #326 (open)
process_ejections -- issue #287 (open)
process_slashings -- issue #322 (open)
process_exit_queue -- issue #323 (open)

File: eth2/beacon/state_machines/forks/serenity/epoch_processing.py [hunk 2]
```diff
@@ -812,76 +821,134 @@ def _check_if_update_validator_registry(state: BeaconState,
...
-        previous_shuffling_start_shard=state.current_shuffling_start_shard,
-        previous_shuffling_seed=state.current_shuffling_seed,
-    )
+StateUpdaterForConfig = Callable[[BeaconState, BeaconConfig], BeaconState]
 
-    need_to_update, num_shards_in_committees = _check_if_update_validator_registry(state, config)
 
-    if need_to_update:
+def _process_validator_registry_with_update(num_shards_in_committees: int) -> StateUpdaterForConfig:
+    def _inner(state: BeaconState, config: BeaconConfig) -> BeaconState:
```
Comment (pipermerriam) [inline eth2/beacon/state_machines/forks/serenity/epoch_processing.py]: IIUC what you are doing here, you could do this without the `inner` function using `toolz.curry` or just a boring `functools.partial`

```python
from eth_utils.toolz import curry

@curry
def _process_validator_registry_with_update(num_shards_in_committees: int, state: BeaconState, config: BeaconConfig) -> BeaconState) -> BeaconState:
    ...

Then you can make the equivalent of your `_inner` function by just calling it with a single argument.

https://toolz.readthedocs.io/en/latest/api.html#toolz.functoolz.curry

File: eth2/beacon/state_machines/forks/serenity/epoch_processing.py [hunk 3]
```diff
@@ -812,76 +821,134 @@ def _check_if_update_validator_registry(state: BeaconState,
...
-            activation_exit_delay=config.ACTIVATION_EXIT_DELAY,
-            latest_active_index_roots_length=config.LATEST_ACTIVE_INDEX_ROOTS_LENGTH,
-            latest_randao_mixes_length=config.LATEST_RANDAO_MIXES_LENGTH,
+        return state
+    return _inner
+
+
+def _process_validator_registry_without_update(state: BeaconState,
+                                               config: BeaconConfig) -> BeaconState:
+    epochs_since_last_registry_change = (
```
Comment (hwwhww) [inline eth2/beacon/state_machines/forks/serenity/epoch_processing.py]: Just found that `epochs_since_last_registry_change` has been renamed to `epochs_since_last_registry_update`.

File: eth2/beacon/state_machines/forks/serenity/epoch_processing.py [hunk 4]
```diff
@@ -812,76 +821,134 @@ def _check_if_update_validator_registry(state: BeaconState,
...
+        return state
+
+    if is_power_of_two(epochs_since_last_registry_change):
+        # Update step-by-step since updated `state.current_shuffling_epoch`
+        # is used to calculate other value). Follow the spec tightly now.
+        state = _update_shuffling_epoch(state, config.SLOTS_PER_EPOCH)
+
+        # NOTE: We do NOT update the "start shard" as we have not
+        # produced a full set of new crosslinks; validators should have a chance to
+        # complete this goal in future epochs.
```
Comment (hwwhww) [inline eth2/beacon/state_machines/forks/serenity/epoch_processing.py]: 👍

File: eth2/beacon/state_machines/forks/serenity/epoch_processing.py (hunk 2 above)
Comment (ralexstokes) [reply eth2/beacon/state_machines/forks/serenity/epoch_processing.py]: yeah this fell out of an intermediate refactor where the vague idea was keeping the `state` as the first parameter....

i agree w/ you that this calls for some `curry`!

**Pull request merged**